### PR TITLE
objstore/swift: use ncw/swift and support large files for OpenStack Swift

### LINKED ISSUE #2678: Issue with large files using Swift storage
While migrating from MinIO to Swift I found out that the Gophercloud library does not support large file support out of the box. Swift has by default limit of 5GB for file (our setup has 500MB).
More info here https://docs.openstack.org/swift/latest/overview_large_objects.html

I'd say this is pretty big limitation since the indexes can grow to enormous file size.

We could implement the Large File support with gophercloud ourselves but it's definitely not trivial.
There is also gophercloud/utils where is helper function for upload which implements the support but it requires the file to be on the disk (regarding out bucket interface it would require changes or writing to tmp file) but this could be possibly changed upstream. Also it has no releases and does not look much maintained. https://github.com/gophercloud/utils/blob/master/openstack/objectstorage/v1/objects/upload.go#L119

Other possibility is to use different library for the Swift.
Probably only one which we could consider to use is https://github.com/ncw/swift
It supports those Large Files so should allow Thanos to upload gig indexes.

@sudhi-vm @bwplotka WDYT?

### PR DESCRIPTION
**Pull request by FUSAKLA**
Resolves #2678 
Closes #2665 

As described in the issue, switch to new library to communicate with swift which supports large file types.

I also changed the maintainer of the swift integration to me since if this would be accepted, there would be not much of the old codebase left so it felt fair not to leave the burden on @sudhi-vm.

It supports also working with the large objects which was limit of the previous library.

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes
- switched to new library for Swift API
- added support for large files (description in docs)
- added timeouts for Swift API 

## Verification

All tests are passing, tried running it as a sidecar, compactor and store for now on real workload.
Works well but will continue testing it with higher load and  bigger blocks. 

Eventually would like to migrate production on it.

### CODE REVIEW

File: CHANGELOG.md
Comment (bwplotka) [inline CHANGELOG.md:27]: Let's add this PR number :hugs:

Comment (bwplotka) [reply CHANGELOG.md:27]: plus link

File: docs/storage.md
Comment (bwplotka) [inline docs/storage.md:307]: wow :hugs:

File: pkg/objstore/swift/swift.go
Comment (bwplotka) [inline pkg/objstore/swift/swift.go:185]:
```suggestion
	} else if _, err := ensureContainer(connection, sc.SegmentContainerName, createContainer); err != nil {
```

File: pkg/objstore/swift/swift.go
Comment (bwplotka) [inline pkg/objstore/swift/swift.go:268]: I would be explicit here (: 

```go
if c.IsObjNotFoundErr(err) {
  return false, nil
}

if err != nil {
  return false, errors.Wrap(err, "swift check if file exists")
}

return true, nil
```

Comment (bwplotka) [reply pkg/objstore/swift/swift.go:268]: bit safer (:

Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:270]: less indentation is always better :+1: :)

**Comment by FUSAKLA** [comment]
thanks for the _swift_ review :grin: 

I'd like to wait some time to test it out more to be sure it works fine. 
Also I'd love to have some other Swift user to test it out but it seems like nobody active is here around :/ 

Regarding the consistency issues. 
I'm aware of Swift eventual consistency but with recently added deletion marks and delay after upload I believe it could work. My biggest concern is probably downsampling of blocks which were recently created by compaction.
But this could be mitigated by separating those by some time span? :thinking: 

Am I missing out some other cases where it would could blow up?

File: CHANGELOG.md
Comment (GiedriusS) [inline CHANGELOG.md:33]: Seems like we are missing a link here.

File: docs/storage.md
Comment (GiedriusS) [inline docs/storage.md:309]: Minor grammar nit: we want a noun here.

```suggestion
The default limit for using SLO is 1 GiB which is also the maximum size of the segment.
```

File: pkg/objstore/swift/swift.go
Comment (GiedriusS) [inline pkg/objstore/swift/swift.go:32]: Maybe it would be worth adding `AuthVersion` here with the value `2`? AFAIK there are only two auth versions available and `2` is the newer one, so adding this default value here would mean that all of the Swift users won't need to add it.

Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:32]: Hm, Actually there are 3 versions, see for example http://doc.isilon.com/ECS/3.1/DataAccessGuide/ecs_c_oapi_swift_authentication_introduction.html

But we could do better with the default value. The library docs states: `Set to 1, 2 or 3 or leave at 0 for autodetect` so `0` might be the way WDYT?

Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:32]: Changed to `0` now

File: CHANGELOG.md
Comment (FUSAKLA) [reply CHANGELOG.md]: Thanos for noticing, added.

**Comment by FUSAKLA** [comment]
Thanks for another review! :) 
The testing looks promising. I see no obvious issues, but unfortunately our cluster has some issues with the load my testing recording rules are generating so there are some issues caused by this from time to time. I'm now waiting for the maintainers of the cluster to tune it up a bit so there are no errors at all and I'm 100% sure it's OK.

I'm fairly confident the uploading/reading/listing etc are working OK.
My biggest concern causing some unexpected breaking change in the configuration of the bucket.
I hope not but I failed to find any previous user of the Swift integration to verify it :confused:

**Comment by FUSAKLA** [comment]
I'll rebase once https://github.com/thanos-io/thanos/pull/2838 to verify the tests still passes even after the refactor

File: pkg/objstore/swift/swift.go
Comment (GiedriusS) [reply pkg/objstore/swift/swift.go:32]: Makes sense, thank you!

**Comment by FUSAKLA** [comment]
I finally found a while to take a look at the failing tests. 
Great idea with running the tests in the CI @GiedriusS , caught two issues actually (one of them regression during CR) :tada: 

Now it's all green so hopefully everything is OK :crossed_fingers: 

I was afraid to merge this right away but it has been a while and it looks like there are no other active users of Swift so I'd go ahead and merge this if you are not against it?

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [inline pkg/objstore/swift/swift.go:219]: @bwplotka @GiedriusS PTAL this feels bit hacky, Swift puts the object segments to the `segments` directory in the configured segments container(bucket). To avoid breaking change and requiring another container, the default segments container is the same as for the original data. This means that in the bucket is directory `segments` full of the file parts. 

In one of the tests we check listing of all objects in the bucket, this is an issue in this case since there is additional object now causing the test to fail. As a solution I added skip of this one particular directory but I'm not sure it this is the right way and it feels bit hacky :/

**Comment by FUSAKLA** [comment]
Rebased once again :recycle:

File: pkg/objstore/swift/swift.go
Comment (GiedriusS) [inline pkg/objstore/swift/swift.go:310]: Is there a reason not to use `io.Copy()`?

File: pkg/objstore/swift/swift.go
Comment (GiedriusS) [inline pkg/objstore/swift/swift.go:316]: I believe this might lead to a very huge buffer. Streaming reading might significantly improve performance. This is related to the previous question.

**Review by GiedriusS — COMMENTED** [review]
Sorry for the delay in the review. Some questions/nits in line.

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: Definitely, should be changed now, thanks makes parsing easier! :+1:

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: done

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: :+1:

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: TBH the PR is quite old and I do not remember but I do not see any obvious reason not to :thinking: 

Changed it to `io.Copy` and will see hot the tests end but definitely good point :)

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: changed to io.copy

**Comment by FUSAKLA** [comment]
@GiedriusS Thanks for going through it in detail. All comments were valid and should be fixed now.

Tests related to Swift are passing but other seems to be flaky probably? :confused: 

PTAL

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:203]: Ok, I'll leave it like it is

**Comment by FUSAKLA** [comment]
@GiedriusS do you think we could move on with this? I'll rebase

**Comment by hsmade** [comment]
We're bumping into the same issue. We can offer to test on our QA envs, but there we might not have big enough buckets to 'need' the splitting up on uploads.

**Comment by kakkoyun** [comment]
@FUSAKLA Could you rebase again? I can help with the review and push this forward.

File: pkg/objstore/swift/swift.go
Comment (GiedriusS) [inline pkg/objstore/swift/swift.go:85]: Very minor thing but in this `Config` we should set `ChunkSize` to `DefaultConfig.ChunkSize`, right? Otherwise it is always zero if `SWIFT_CHUNK_SIZE` is empty.

File: pkg/objstore/swift/swift.go
Comment (GiedriusS) [inline pkg/objstore/swift/swift.go:141]:
```suggestion
		return nil, errors.Wrap(err, "swift parse config")
```

File: pkg/objstore/swift/swift.go
Comment (GiedriusS) [inline pkg/objstore/swift/swift.go:216]:
```suggestion
		return nil, errors.New("object name cannot be empty")
```

File: pkg/objstore/swift/swift.go
Comment (GiedriusS) [inline pkg/objstore/swift/swift.go:242]:
```suggestion
		return objstore.ObjectAttributes{}, errors.New("object name cannot be empty")
```

File: pkg/objstore/swift/swift.go
Comment (GiedriusS) [inline pkg/objstore/swift/swift.go:295]:
```suggestion
		return errors.Wrap(err, "downloading swift object")
```

File: pkg/objstore/swift/swift.go
Comment (GiedriusS) [inline pkg/objstore/swift/swift.go:347]: What's the reason for this? :P

**Review by GiedriusS — COMMENTED** [review]
Very sorry for the long delay in reviewing this. Looks good, just some minor nits.

**Comment by FUSAKLA** [comment]
Ill adress all the comments ASAP, can't wait to see this merged :smile:

File: pkg/objstore/swift/swift.go
Comment (deniszh) [inline pkg/objstore/swift/swift.go:280]: Could we optionally do this with `DynamicLargeObjectCreateFile` too (with some switch and SLO by default)? I know that SLO is used here for consistency, but e.g. our Swift do not support SLO but support DLO.
Much appreciated with effort BTW!

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:85]: Right, I missed that, thanks!

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:295]: hm, there is uploading going on? so maybe `uploading swift object` ?  but otherwise :+1: from me for the suggestion

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:347]: :neutral_face: TBH I do not remember after that time :sweat_smile: 

I'll investigate, maybe there was an issue with the eventual consistency for tests :/

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:280]: Hi, I suppose it won't hurt to add the support for DLO, but I'll add warning to the docs about it's usage and consequences to scare off the users :)

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: fixed in https://github.com/thanos-io/thanos/pull/2732/commits/7a51f585a7184cc9a0c5a73ffcd77b23896da026

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: resolved in https://github.com/thanos-io/thanos/pull/2732/commits/7a51f585a7184cc9a0c5a73ffcd77b23896da026

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: Added here https://github.com/thanos-io/thanos/pull/2732/commits/7a51f585a7184cc9a0c5a73ffcd77b23896da026

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:311]: changed here https://github.com/thanos-io/thanos/pull/2732/commits/7a51f585a7184cc9a0c5a73ffcd77b23896da026 to `uploading swift object`, hope that's ok

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: The DLO support is added here https://github.com/thanos-io/thanos/commit/7a51f585a7184cc9a0c5a73ffcd77b23896da026, check it out.

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [inline pkg/objstore/swift/swift.go:46]: Hm, we managed to get to an interesting situation.. 

Cortex stared to use the old `"github.com/thanos-io/thanos/pkg/objstore/swift".SwiftConfig` which I'm renaming to just `Config`.
But we are obviously embedding Cortex code also so this leads to issue that I'm not able to build id.  

This workaround adds the `SwiftConfig` type just to make it working but I'm not sure if this is a way to go.
Or should I leave the Name of the config to `SwiftConfig`?

 @GiedriusS I'd be really glad for any advice here

**Comment by FUSAKLA** [comment]
Not sure if the e2e test fail is related, does not look like it

**Comment by deniszh** [comment]
Btw, we have agreement in
https://github.com/cortexproject/cortex/pull/3303 to fix Swift support in
Cortex when this PR will be merged.

**Comment by FUSAKLA** [comment]
Cool! thanks for info :tada: :)

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: The tests passes even with the DLO as I tested it now so hopefully it will work out for you :)

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: Tested it without the timeout and the tests are passing so I'll just drop it :)
Thanks for noticing it.

**Comment by FUSAKLA** [comment]
Thanks @GiedriusS! 
PTAL once again hopefully the last round :)

File: pkg/objstore/swift/swift.go
Comment (bwplotka) [inline pkg/objstore/swift/swift.go:18]: formatting is needed, but that's fine

File: pkg/objstore/swift/swift.go
Comment (bwplotka) [inline pkg/objstore/swift/swift.go:34]: Can we comment? What's this?

File: pkg/objstore/swift/swift.go
Comment (bwplotka) [reply pkg/objstore/swift/swift.go:46]: There is a semi-good way to handle this. I will respond with some blog post I think. (: For now, this is fine :+1:

Comment (bwplotka) [inline pkg/objstore/swift/swift.go:46]:
```suggestion
type SwiftConfig = Config
```
This makes it clear that it's an alias

File: pkg/objstore/swift/swift.go
Comment (bwplotka) [inline pkg/objstore/swift/swift.go:175]: Normally we don't need to add `swift` to wrap. In the same way, we don't need to add `NewContainerFromConfig swift...`. This is because in Thanos we wrap "whatever we do now" with context, not "where we do it"

File: pkg/objstore/swift/swift.go
Comment (bwplotka) [inline pkg/objstore/swift/swift.go:266]: None of those support context cancellation?

File: pkg/objstore/swift/swift.go
Comment (bwplotka) [reply pkg/objstore/swift/swift.go:311]: still not changed here though

File: pkg/objstore/swift/swift.go
Comment (bwplotka) [inline pkg/objstore/swift/swift.go:307]:
```suggestion
		return errors.Wrap(err, "create file")
```

Comment (bwplotka) [reply pkg/objstore/swift/swift.go:307]: Maybe it would nice to get those near each create method so we know if it's statc, dynamic or objectCreate

**Review by bwplotka — APPROVED** [review]
LGTM, just small nits.

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:46]: Great, thanks! I'd be happy to read about that :)

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:263]: Unfortunately not. The connection lets you pass own transport, but that does not help :/

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: Moved them right to the creating of the file so the wrap can distinguish SLO, DLO and normal object

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: Right, I removed the swift reference from all the error wrappings.

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go:31]: Comment added, hopefully it's understandable

File: pkg/objstore/swift/swift.go
Comment (FUSAKLA) [reply pkg/objstore/swift/swift.go]: sorry, should be fixed now :crossed_fingers:

**Comment by FUSAKLA** [comment]
Thanks for looking at it once again! :+1: 
I hope I addressed all that I could.

Regarding the context cancellation.. I'm not sure we can do much about it here.

**Comment by FUSAKLA** [comment]
@bwplotka @GiedriusS PTAL, should be up to date again
e2e tests not sure why :/

**Review by GiedriusS — APPROVED** [review]
The code looks good but I'm kind of hesitant to merge this because the new library doesn't support `context.Context` and that means we have to add yet another timeout parameter. @FUSAKLA perhaps you know if there are any plans to add support for contexts to the library?

**Comment by FUSAKLA** [comment]
@GiedriusS Yep, that is not great but the current state of the swift integration is almost unusable because of the file size limitation.

I opened an issue on the library and it showed up it's already considered (it was originally created before Google context library was created :)) ) so I'd send a PR there to add it.

Considering the current status of the Swift integration and how long is this PR pending, I'd be glad to merge this and ASAP send PR to add the context support once done in the library if you wouldn't mind. Swift is still marked as experimental in the docs so maybe the lack of context is acceptable? 

See the issues: 
https://github.com/ncw/swift/issues/161
https://github.com/ncw/swift/issues/159

**Comment by bwplotka** [comment]
Happy with this approach given the inusability of previous integration. Thanks for this work and sorry for lag in review.

Good work!

**Pull request merged**
